user_status: Live update user status icon in recent conversations.

kuv2707 opened the pull request:
Currently, the user circles indicating presence are not synced across the frontend. In this PR, we sync them across:
- Inbox row
- Recent view row
- User card popover
- User profile modal

Fixes #30536

**Screen capture**

*After few moments of inactivity, when the presence status of ZOE is updated to blank circle, all the circles turn blank*
*Likewise, upon logging ZOE in again, all circles turn green*

https://github.com/zulip/zulip/assets/88523649/91264c7d-151f-4e6d-8c50-2e8d2197990e




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

kuv2707 commented:
@pratik-pc Please have a look

kuv2707 commented:
@zulipbot add "buddy review"

pratik-pc commented:
@zulipbot remove "buddy review"

kuv2707 commented:
@zulipbot add "mentor review"

Riken-Shah commented:
Looks pretty straightforward, should we add a test to verify the state switch?

kuv2707 commented:
@amanagr Do we need to add tests for this?

timabbott commented on web/src/activity_ui.ts:
```diff
@@ -59,6 +59,18 @@ export function clear_for_testing(): void {
+        if (Number.isNaN(user_id)) {
+            // This eliminates elements having more than one user-id.
+            return;
+        }
```
There shouldn't be any, right? So you might want an assert here.

kuv2707 replied:
Done

timabbott commented on web/src/activity_ui.ts:
```diff
@@ -59,6 +59,18 @@ export function clear_for_testing(): void {
 
+function update_status_circles(): void {
+    $(".user_circle").each(function () {
+        const user_id = Number.parseInt($(this).attr("data-status-sync-user-id") ?? "", 10);
```
Let's use "presence" rather than "status" here; this is about presence/availability, not the "available" status.

`data-presence-indicator-user-id` and `update_presence_indicators()` might be appropriate.

kuv2707 replied:
Done

timabbott commented on web/tests/lib/zjquery_element.js:
```diff
@@ -62,6 +62,10 @@ function FakeElement(selector, opts) {
+        each(cb) {
+            cb.call($self);
+            return $self;
+        },
```
This doesn't look like a loop at all? Maybe I'm missing something.

I don't think you need tests specifically for this block of code -- manual testing should be sufficient -- but let's try not to add a test helper function that doesn't behave as described.

kuv2707 replied:
I think the function `update_presence_indicators` should be stubbed out.

alya commented:
Tagging as "integration review", since @timabbott is currently reviewing this, and @Riken-Shah has commented.

alya commented:
Did this PR end up fixing all the additional issues noted in this comment https://github.com/zulip/zulip/issues/30536#issuecomment-2198234122 ?

kuv2707 commented:
Not yet. I wasn't sure if we want to fix them since those popovers are short-lived.

The fix only involves adding a classname though, so its a small effort.

kuv2707 commented on web/tests/activity.test.js:
```diff
@@ -537,6 +537,16 @@ test("insert_one_user_into_empty_list", ({override, mock_template}) => {
         $other_users_appended = $element;
     });
 
+    $.create("[data-presence-indicator-user-id]", {
```
This has been done to restore full test coverage.

kuv2707 commented on web/tests/activity.test.js:
```diff
@@ -537,6 +537,16 @@ test("insert_one_user_into_empty_list", ({override, mock_template}) => {
         $other_users_appended = $element;
     });
 
+    $.create("[data-presence-indicator-user-id]", {
+        children: [
+            {
+                dataset: {
+                    presenceIndicatorUserId: 1,
+                },
+            },
+        ],
+    });
+
```
This has been done to restore full test coverage.
(Used by `activity_ui.update_presence_indicators`)

kuv2707 commented:
Made the following changes:

- Removed the `.each` stub from zjquery and used `override_rewire` instead.
- Addressed [this comment](https://github.com/zulip/zulip/issues/30536#issuecomment-2198234122) (syncing most user presence indicators)

timabbott commented on web/src/activity_ui.ts:
```diff
@@ -59,6 +59,18 @@ export function clear_for_testing(): void {
+        this.classList.remove("user_circle_empty");
+        this.classList.remove("user_circle_green");
+        this.classList.remove("user_circle_idle");
+        this.classList.add(user_circle_class);
```
This should use jQuery, and in particular `.attr("data-presence-indicator-user-id")`, like the rest of the codebase.

(`this.dataSet` reformatting the data attribute name is really annoying for `git grep`).

kuv2707 replied:
@timabbott Changed to jQuery

timabbott commented:
Merged, thanks @kuv2707!
